src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java

```diff
@@ -175,6 +175,35 @@ public boolean isReady(@Nonnull JCloudsSlave slave) {
         public abstract boolean isReady(@Nonnull JCloudsSlave slave);
     }
 
+    public enum BootSource {
```

Comment:
I'm sure this is possible, but the only existing pattern in the code is `JCloudsCloud.SlaveType`, and this has a "TODO: refactor to interface/extension point" comment in front of it, which is why I didn't do it that way originally.

I could follow the lead of `JCloudsCloud.SlaveType` and embed the functionality into the `BootSource` enum itself.  If that's the sort of thing you're after then I think I'm fairly clear on what that entails.
e.g.
`public enum BootSource {
IMAGE { ... },
VOLUMESNAPSHOT { ... };
/** For provisioning */
public abstract void addSource(ServerCreateBuilder builder, Openstack openstack);
/** For Web form */
public abstract boolean populateImageIds(Openstack openstack, ListBoxModel m, String existingValue);
/** For Web form validation */
public abstract int countMatchingImages(Openstack openstack, String imageId);
}`

If that's not what you want then I'll need a bit more guidance in order to avoid just creating more code that isn't what you want :-)  Can you point me at some existing Jenkins/plugin code that I can use as an example?